crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1101 @@
+            }
+        }
+    }
+}
+
+fn reversed_synth_su4_sequence(
+    mut su4_mat: Array2<Complex64>,
+    decomposer_2q: &DecomposerElement,
+    approximation_degree: Option<f64>,
+) -> PyResult<TwoQubitUnitarySequence> {
+    let is_approximate = approximation_degree.is_none() || approximation_degree.unwrap() != 1.0;
+    // Swap rows 1 and 2
+    let (mut row_1, mut row_2) = su4_mat.multi_slice_mut((s![1, ..], s![2, ..]));
+    azip!((x in &mut row_1, y in &mut row_2) (*x, *y) = (*y, *x));
+
+    // Swap columns 1 and 2
+    let (mut col_1, mut col_2) = su4_mat.multi_slice_mut((s![.., 1], s![.., 2]));
+    azip!((x in &mut col_1, y in &mut col_2) (*x, *y) = (*y, *x));
+
+    let synth = if let DecomposerType::TwoQubitBasisDecomposer(decomp) = &decomposer_2q.decomposer {
+        decomp.call_inner(su4_mat.view(), None, is_approximate, None)?
+    } else {
+        panic!("reversed_synth_su4_sequence should only be called for TwoQubitBasisDecomposer.")
+    };
+
+    let flip_bits: [u8; 2] = [1, 0];
+    let mut reversed_gates = Vec::with_capacity(synth.gates().len());
+    for (gate, params, qubit_ids) in synth.gates() {
+        let new_qubit_ids = qubit_ids
+            .into_iter()
+            .map(|x| flip_bits[*x as usize])
+            .collect::<SmallVec<[u8; 2]>>();
+        reversed_gates.push((*gate, params.clone(), new_qubit_ids.clone()));
+    }
+
+    let mut reversed_synth: TwoQubitGateSequence = TwoQubitGateSequence::new();
+    reversed_synth.set_state((reversed_gates, synth.global_phase()));
+    let sequence = TwoQubitUnitarySequence {
+        gate_sequence: reversed_synth,
+        decomp_gate: decomposer_2q.gate.clone(),
+    };
+    Ok(sequence)
+}
+
+fn synth_su4_dag(
+    py: Python,
+    su4_mat: &Array2<Complex64>,
+    decomposer_2q: &DecomposerElement,
+    preferred_direction: Option<bool>,
+    approximation_degree: Option<f64>,
+) -> PyResult<DAGCircuit> {
+    let is_approximate = approximation_degree.is_none() || approximation_degree.unwrap() != 1.0;
+    let synth_dag = if let DecomposerType::XXDecomposer(decomposer) = &decomposer_2q.decomposer {
+        let kwargs: HashMap<&str, bool> = [("approximate", is_approximate), ("use_dag", true)]
+            .into_iter()
+            .collect();
+        decomposer
+            .call_bound(
+                py,
+                (su4_mat.clone().into_pyarray_bound(py),),
+                Some(&kwargs.into_py_dict_bound(py)),
+            )?
+            .extract::<DAGCircuit>(py)?
+    } else {
+        panic!("synth_su4_dag should only be called for XXDecomposer.")
+    };
+
+    match preferred_direction {
+        None => Ok(synth_dag),
+        Some(preferred_dir) => {
+            let mut synth_direction: Option<Vec<u32>> = None;
+            for node in synth_dag.topological_op_nodes()? {
+                let NodeType::Operation(inst) = &synth_dag.dag()[node] else {
+                    panic!("DAG node must be an instruction")
+                };
```

Comment:
Could reduce a couple lines when using `unwrap_operation`.
```suggestion
                let inst = &synth_dag.dag()[node].unwrap_operation();
```